pennylane/capture/measure.py

```diff
@@ -0,0 +1,134 @@
+# Copyright 2024 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This submodule defines the measure primitive for performing measurements in plxpr.
+"""
+from functools import lru_cache
+
+import pennylane as qml  # need shots class without circular dependency issues
+
+has_jax = True
+try:
+    import jax
+except ImportError:
+    has_jax = False
+
+
+@lru_cache
+def _get_measure_primitive():
+    if not has_jax:
+        raise ImportError("jax is required to create the measure primitive.")
+
+    if jax.config.jax_enable_x64:
+        dtype_map = {
+            float: jax.numpy.float64,
+            int: jax.numpy.int64,
+            complex: jax.numpy.complex128,
+        }
+    else:
+        dtype_map = {
+            float: jax.numpy.float32,
+            int: jax.numpy.int32,
+            complex: jax.numpy.complex64,
+        }
+
+    measure_prim = jax.core.Primitive("measure")
+    measure_prim.multiple_results = True
+
+    # def trivial_processing(*results):
+    #    return results
+
+    # pylint: disable=unused-argument
+    @measure_prim.def_impl
+    def _(*measurements, shots, num_device_wires):
+        # depends on the jax interpreter
+        if not all(isinstance(m, qml.measurements.MidMeasureMP) for m in measurements):
+            raise NotImplementedError("requires an interpreter to perform a measurement.")
+        raise NotImplementedError("currently causes jax to enter an infinite loop")
+        # TODO: figure out how to make this return a measurement value
+        # return qml.measurements.MeasurementValue(measurements, trivial_processing)
```

Comment:
I suppose this is out of scope, so these two lines could be updated either by removing or by providing come context of what this implementation should be doing exactly once we build it :)

Reply:
I was trying to turn something like `qml.measure(0)` back from jaxpr to `eval_jaxpr` qfunc behavior, it would just hang.  I'm betting because a `MeasurementValue` isn't actually an array of shape `()` and type `int32`.

Not sure how to solve this problem yet.  Potentially we could make `MeasurementValue` inherit from numpy array and give it a shape and dtype?

Reply:
Ah I see! Yeah that sounds nice, although it seems like something we want to leave to a dedicated MCM support PR, maybe? :)